src/http/server.cr

```diff
@@ -95,90 +100,140 @@ class HTTP::Server
+  def bind_tcp(port : Int32, reuse_port : Bool = false) : Socket::IPAddress
+    bind_tcp "127.0.0.1", port, reuse_port
   end
 
-  # Returns the TCP port the server is connected to.
+  # Creates a `TCPServer` listening on an unused port and adds it as a socket.
+  #
+  # Returns the `Socket::IPAddress` with the determined port number.
   #
-  # For example you may let the system choose a port, then report it:
   # ```
-  # server = HTTP::Server.new(0) { }
-  # server.bind
-  # server.port # => 12345
+  # server = HTTP::Server.new { }
+  # server.bind_unused_port # => Socket::IPAddress.new("127.0.0.1", 12345)
   # ```
-  def port
-    if server = @server
-      server.local_address.port.to_i
-    else
-      @port
+  def bind_unused_port(host : String = "127.0.0.1", reuse_port : Bool = false) : Socket::IPAddress
+    bind_tcp host, 0, reuse_port
+  end
+
+  # Adds a `Socket::Server` *socket* to this server.
+  def bind(socket : Socket::Server) : Socket::Server
+    @sockets << socket
+
+    socket
+  end
+
+  # Enumerates all addresses this server is bound to.
+  def each_address(&block : Socket::Address ->)
+    @sockets.each do |socket|
+      yield socket.local_address
     end
   end
 
-  # Creates the underlying `TCPServer` if the doesn't already exist.
+  def addresses : Array(Socket::Address)
+    array = [] of Socket::Address
+    each_address do |address|
+      array << address
+    end
+    array
+  end
+
+  # Creates a `TCPServer` listenting on `127.0.0.1:port`, adds it as a socket
+  # and starts the server. Blocks until the server is closed.
   #
-  # You may set *reuse_port* to true to enable the `SO_REUSEPORT` socket option,
-  # which allows multiple processes to bind to the same port.
-  def bind(reuse_port = false)
-    @server ||= TCPServer.new(@host, @port, reuse_port: reuse_port)
+  # See `#bind(port : Int32)` for details.
+  def listen(port : Int32, reuse_port : Bool = false)
+    bind_tcp(port, reuse_port)
+
+    listen
   end
 
-  # Starts the server. Blocks until the server is closed.
+  # Creates a `TCPServer` listenting on `host:port`, adds it as a socket
+  # and starts the server. Blocks until the server is closed.
   #
-  # See `#bind` for details on the *reuse_port* argument.
-  def listen(reuse_port = false)
-    server = bind(reuse_port)
-    until @wants_close
-      spawn handle_client(server.accept?)
+  # See `#bind(host : String, port : Int32)` for details.
+  def listen(host : String, port : Int32, reuse_port : Bool = false)
+    bind_tcp(host, port, reuse_port)
+
+    listen
+  end
+
+  # Starts the server. Blocks until the server is closed.
+  def listen
+    raise "Can't start server with not sockets to listen to" if @sockets.empty?
+
+    done = Channel(Nil).new
+
+    @sockets.each do |socket|
+      spawn do
+        until @wants_close
+          spawn handle_client(socket.accept? || break)
+        end
+
+        done.send nil
+      end
     end
+
+    @sockets.size.times { done.receive }
```

Comment:
What if `#accept?` raises? It can raise an `Errno` for example.

Reply:
Should be fixed now.